### PR TITLE
Fix `filter_var()` narrowing with unknown options

### PR DESCRIPTION
**Pull request by herndlm**
Closes https://github.com/phpstan/phpstan/issues/8516

started the filter-var.php node scope resolver test file because I'd like to move some other filter-var tests over there

also: the fix is the first commit, the second one is still related, but the rest are unrelated fixes and improvements in additional commits.. thought I'll save you from 10 more PRs here :)

### CODE REVIEW

**Comment by herndlm** [comment]
ah ok, the failing GenericsIntegrationTest / LevelsTest case is unrelated, exists also in other PRs

File: tests/PHPStan/Analyser/data/filter-var-returns-non-empty-string.php [hunk 1]
```diff
@@ -82,7 +82,7 @@ public function run(string $str, int $int, int $positive_int, int $negative_int)
 		assertType('9', $return);
 
 		$return = filter_var(1.0, FILTER_VALIDATE_INT, ['options' => ['min_range' => 1, 'max_range' => 9]]);
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/filter-var-returns-non-empty-string.php:84]: I guess you could now also test/assert `filter_var(111.0, FILTER_VALIDATE_INT, ['options' => ['min_range' => 1, 'max_range' => 9]])` that it returns `false`?

Comment (herndlm) [reply tests/PHPStan/Analyser/data/filter-var-returns-non-empty-string.php:84]: good idea, it does luckily 😅 I do not like this extension..

File: tests/PHPStan/Analyser/data/filter-var.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+		assertType('array<false>', filter_var(false, FILTER_VALIDATE_BOOLEAN, FILTER_FORCE_ARRAY | FILTER_NULL_ON_FAILURE));
+
+		assertType('17', filter_var(17, FILTER_VALIDATE_INT));
+
+		assertType('17.0', filter_var(17, FILTER_VALIDATE_FLOAT));
+
+		assertType("'17'", filter_var(17));
+		assertType("'17'", filter_var(17.0));
+		assertType("'1'", filter_var(true));
+		assertType("''", filter_var(false));
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/filter-var.php:85]: Does this also mean?

```suggestion
		assertType("'1'", filter_var(true));
		assertType("''", filter_var(false));
     assertType("''|'1'", filter_var($bool));

```

Comment (herndlm) [reply tests/PHPStan/Analyser/data/filter-var.php:85]: Yes, it does, this is correctly implemented in BoolType already. Will add more tests tomorrow, thx

**Comment by herndlm** [comment]
sorry for the pressure, but what do you think about this one? it's big I know.. afterwards I should be finally able to re-use the logic and add support for `filter_input()` in the other PR. and I'm looking forward very much to that one, to help with many errors in some legacy code base.. 😊

**Comment by ondrejmirtes** [comment]
This is so complicated 😂 Thank you.

**Pull request merged**